### PR TITLE
Introducing .NET version of css-layout using P/Invoke

### PR DESCRIPTION
**Pull request by mattpodwysocki**
This version of the css-layout project includes support for .NET projects.  Up in the air is how many configurations of .NET projects we allow for, such as Portable Class Libraries, Universal Windows Platform, .NET Core, etc.  Still needs integration with Buck.

### CODE REVIEW

**Comment by facebook-github-bot-8** [comment]
Thank you for your pull request and welcome to our community. We require contributors to sign our Contributor License Agreement, and we don't seem to have you on file. In order for us to review and merge your code, please sign up at https://code.facebook.com/cla - and if you have received this in error or have any questions, please drop us a line at cla@fb.com. Thanks!

**Comment by facebook-github-bot-8** [comment]
@mattpodwysocki updated the pull request - [view changes](https://github.com/facebook/css-layout/pull/220/files/31befa176c9ce2c42161c7451118062a38d39f74..678d2284f7d172b7da55041a4fc7258451bcf57d)

File: csharp/.gitignore [hunk 1]
```diff
@@ -0,0 +1,265 @@
+## Ignore Visual Studio temporary files, build results, and
```
Comment (emilsjolander) [inline csharp/.gitignore]: please mirror this in a corresponding .hgignore

File: .gitignore [hunk 2]
```diff
@@ -6,3 +6,6 @@
 /.buckd
 /lib/gtest/googletest-*/
 /gentest/test.html
+
+# Visual studio code
```
Comment (emilsjolander) [inline .gitignore]: Please mirror this in .hgignore

File: csharp/CSSLayout/CSSLayout-win.h [hunk 3]
```diff
@@ -0,0 +1,206 @@
+/**
```
Comment (emilsjolander) [inline csharp/CSSLayout/CSSLayout-win.h]: This seems annoying to maintain. Can we make this use the same .h file as the C version?

File: csharp/Facebook.CSSLayout/CSSCachedMeasurement.cs [hunk 4]
```diff
@@ -0,0 +1,12 @@
+﻿namespace Facebook.CSSLayout
+{
+    public class CSSCachedMeasurement
```
Comment (emilsjolander) [inline csharp/Facebook.CSSLayout/CSSCachedMeasurement.cs]: This class should not be needed as it is not part of the public api

File: csharp/Facebook.CSSLayout/CSSLayout.cs [hunk 5]
```diff
@@ -0,0 +1,66 @@
+﻿using System.Runtime.InteropServices;
+
+namespace Facebook.CSSLayout
+{
+    public class CSSLayout
+    {
```
Comment (emilsjolander) [inline csharp/Facebook.CSSLayout/CSSLayout.cs]: This class should not be needed as it is handled in the C code.

File: csharp/Facebook.CSSLayout/CSSLayoutContext.cs [hunk 6]
```diff
@@ -0,0 +1,8 @@
+﻿namespace Facebook.CSSLayout
+{
+    public class CSSLayoutContext
```
Comment (emilsjolander) [inline csharp/Facebook.CSSLayout/CSSLayoutContext.cs]: This class should not be needed as the C version does not have it.

File: csharp/.gitignore (hunk 1 above)
Comment (mattpodwysocki) [inline csharp/.gitignore]: @emilsjolander ok, will do

File: csharp/CSSLayout/CSSLayout-win.h (hunk 3 above)
Comment (mattpodwysocki) [inline csharp/CSSLayout/CSSLayout-win.h]: @emilsjolander we could if you want pragma comments for checking if Windows, sure

File: csharp/Facebook.CSSLayout/CSSLayoutContext.cs (hunk 6 above)
Comment (mattpodwysocki) [inline csharp/Facebook.CSSLayout/CSSLayoutContext.cs]: @emilsjolander is the intent then not to implement the regular `CSSNode` class and only implement the native class?

File: csharp/CSSLayout/CSSLayout-win.h (hunk 3 above)
Comment (mattpodwysocki) [inline csharp/CSSLayout/CSSLayout-win.h]: @emilsjolander should we update it as such? https://gist.github.com/mattpodwysocki/c48da3dd668708ee227263b2ae382815

**Comment by facebook-github-bot-6** [comment]
@mattpodwysocki updated the pull request - [view changes](https://github.com/facebook/css-layout/pull/220/files/678d2284f7d172b7da55041a4fc7258451bcf57d..e3b1710b68d1aec5e2169a5a04641fa0bc1fc341)

File: csharp/CSSLayout/CSSLayout-win.h (hunk 3 above)
Comment (emilsjolander) [inline csharp/CSSLayout/CSSLayout-win.h]: perfect!

File: csharp/Facebook.CSSLayout/CSSLayoutContext.cs (hunk 6 above)
Comment (emilsjolander) [inline csharp/Facebook.CSSLayout/CSSLayoutContext.cs]: yup!

File: csharp/.gitignore (hunk 1 above)
Comment (mattpodwysocki) [inline csharp/.gitignore]: @emilsjolander done

File: .gitignore (hunk 2 above)
Comment (mattpodwysocki) [inline .gitignore]: @emilsjolander done

File: csharp/CSSLayout/CSSLayout-win.h (hunk 3 above)
Comment (mattpodwysocki) [inline csharp/CSSLayout/CSSLayout-win.h]: @emilsjolander removed as per request

File: csharp/Facebook.CSSLayout/CSSCachedMeasurement.cs (hunk 4 above)
Comment (mattpodwysocki) [inline csharp/Facebook.CSSLayout/CSSCachedMeasurement.cs]: @emilsjolander removed as per request

File: csharp/Facebook.CSSLayout/CSSLayoutContext.cs (hunk 6 above)
Comment (mattpodwysocki) [inline csharp/Facebook.CSSLayout/CSSLayoutContext.cs]: @emilsjolander removed as per request

**Comment by facebook-github-bot-0** [comment]
Thank you for signing our Contributor License Agreement. We can now accept your code for this (and any) Facebook open source project. Thanks!

**Comment by emilsjolander** [comment]
Everything is looking great. Thanks for being so responsive! Only thing I see missing is integrating this with buck. You can find more info on building csharp with buck here https://buckbuild.com/rule/csharp_library.html

**Comment by facebook-github-bot-0** [comment]
@mattpodwysocki updated the pull request - [view changes](https://github.com/facebook/css-layout/pull/220/files/df024efa07a50b8c26a5a2b94154c4cccc601a01..ef0639e94b07d858776355e6ca2124c056342d9d)

**Comment by emilsjolander** [comment]
I chatted with @mattpodwysocki and there are a couple things that are blocking proper buck support for this so we will skip getting this working on this pull request and instead we plan on fixing this afterwards.

**Comment by facebook-github-bot-1** [comment]
@mattpodwysocki updated the pull request - [view changes](https://github.com/facebook/css-layout/pull/220/files/ef0639e94b07d858776355e6ca2124c056342d9d..2841da4958b49d35209c714d59bae745c64061fa)

**Comment by emilsjolander** [comment]
Looking great! Thanks for the super quick response times. I'll try to merge this today.

**Comment by facebook-github-bot-1** [comment]
Thanks for importing. If you are a Facebook employee, you can view this diff [on Phabricator](https://phabricator.intern.facebook.com/D3909367).

**Comment by langpavel** [comment]
Can you clarify goodness? It's nothing in code, only confused me more..

**Comment by emilsjolander** [comment]
Pavel, what are you referring to?

File: csharp/.hgignore [hunk 7]
```diff
@@ -0,0 +1,265 @@
+## Ignore Visual Studio temporary files, build results, and
```
Comment (Daniel15) [inline csharp/.hgignore]: I wonder if `.hgignore` could just be a symlink to `.gitignore`?

File: csharp/Facebook.CSSLayout/CSSLayoutContext.cs [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ */
+
+namespace Facebook.CSSLayout
+{
+    public class CSSLayoutContext
+    {
+        public MeasureOutput measureOutput { get; } = new MeasureOutput();
```
Comment (Daniel15) [inline csharp/Facebook.CSSLayout/CSSLayoutContext.cs]: Uppercase first character of property name (`measureOutput` -> `MeasureOutput`)?

File: csharp/Facebook.CSSLayout/CSSNode.cs [hunk 9]
```diff
@@ -0,0 +1,685 @@
...
+        {
+            _measureFunc = MeasureInternal;
+            _printFunc = PrintInternal;
+        }
+
+        private void CheckDisposed()
+        {
+            if (_isDisposed)
+            {
+                throw new ObjectDisposedException("CSSNode");
```
Comment (Daniel15) [inline csharp/Facebook.CSSLayout/CSSNode.cs]: `nameof(CSSNode)` assuming you're using C# 6.0 :smile:

File: csharp/Facebook.CSSLayout/CSSNode.cs [hunk 10]
```diff
@@ -0,0 +1,685 @@
...
+                CheckDisposed();
+                return Native.CSSNodeStyleGetFlex(_cssNode);
+            }
+
+            set
+            {
+                CheckDisposed();
+                Native.CSSNodeStyleSetFlex(_cssNode, value);
+            }
+        }
```
Comment (Daniel15) [inline csharp/Facebook.CSSLayout/CSSNode.cs]: All these properties feel like a lot of boilerplate, I wonder if they could be autogenerated?

File: csharp/Facebook.CSSLayout/Native.cs [hunk 11]
```diff
@@ -0,0 +1,263 @@
...
+using System;
+using System.Runtime.InteropServices;
+
+namespace Facebook.CSSLayout
+{
+    internal static class Native
+    {
+        private const string DllName = "CSSLayout.dll";
+
+        [DllImport(DllName)]
```
Comment (Daniel15) [inline csharp/Facebook.CSSLayout/Native.cs]: I think this should omit the `.dll` suffix so it could potentially work on Linux in the future.

**Comment by Daniel15** [comment]
The C stuff looks a bit Windows-specific; it would be good to get a build working on Linux too so this could be used on .NET Core or on Mono :smile:

File: csharp/.hgignore (hunk 7 above)
Comment (emilsjolander) [inline csharp/.hgignore]: Yeah probably

**Comment by emilsjolander** [comment]
@Daniel15 Feel free to open a pull request with further platform support 👍

**Pull request closed without merging**